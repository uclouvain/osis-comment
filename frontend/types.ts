import type {CommentLinks, EntryRecord} from "./interfaces";

export class Entry {
  uuid: string;
  comment: string;
  author: string | null;
  tags: string[];
  extra_data: object;
  links: CommentLinks;

  created_at: Date;

  constructor(
      {
        uuid,
        comment,
        author,
        tags,
        extra_data,
        links,
        created_at,
      }: EntryRecord,
  ) {
    this.uuid = uuid;
    this.comment = comment;
    this.author = author;
    this.tags = tags;
    this.extra_data = extra_data;
    this.links = links;
    this.created_at = new Date(created_at);
  }
}
