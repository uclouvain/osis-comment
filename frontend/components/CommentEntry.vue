<!--
  -
  - OSIS stands for Open Student Information System. It's an application
  - designed to manage the core business of higher education institutions,
  - such as universities, faculties, institutes and professional schools.
  - The core business involves the administration of students, teachers,
  - courses, programs and so on.
  -
  - Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
  -
  - This program is free software: you can redistribute it and/or modify
  - it under the terms of the GNU General Public License as published by
  - the Free Software Foundation, either version 3 of the License, or
  - (at your option) any later version.
  -
  - This program is distributed in the hope that it will be useful,
  - but WITHOUT ANY WARRANTY; without even the implied warranty of
  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  - GNU General Public License for more details.
  -
  - A copy of this license - GNU General Public License - is available
  - at the root of the source code of this program.  If not,
  - see http://www.gnu.org/licenses/.
  -
  -->

<template>
  <div class="panel-body clearfix">
    <div class="comment-authoring">
      <strong>{{ entry.author || $t('entry.anonymous') }}</strong>
      {{
        $t('entry.authored_date', {
          date: entry.created_at.toLocaleDateString(),
          time: entry.created_at.toLocaleTimeString(),
        })
      }}
    </div>
    <template v-if="!isEditing">
      <div class="pull-right comment-actions">
        <button
            v-if="typeof entry.links.edit === 'string'"
            class="btn btn-sm btn-default"
            :title="$t('entry.edit_comment')"
            @click="isEditing = true"
        >
          <i class="fas fa-pencil" />
        </button>
        <button
            v-if="typeof entry.links.delete === 'string'"
            class="btn btn-sm btn-danger"
            :title="$t('entry.delete_comment')"
            @click="$emit('delete', entry.links.delete)"
        >
          <i class="fas fa-trash-alt" />
        </button>
      </div>
      <div
          v-if="!richTextConfig"
          class="comment-content"
      >
        <!-- eslint-disable vue/no-v-html -->
        <p v-html="entry.comment.replace(/\n/g, '<br/>')" />
      </div>
      <!-- eslint-disable vue/no-v-html -->
      <div
          v-else
          class="comment-content"
          v-html="entry.comment"
      />
    </template>
    <CommentEditor
        v-else
        :default-value="entry.comment"
        :rich-text-config="richTextConfig"
        @submit="(value) => $emit('edit', entry.links.edit, value)"
        @cancel="isEditing = false"
    />
  </div>
</template>

<script lang="ts">
import {Entry} from '../types';
import CommentEditor from './CommentEditor.vue';
import {defineComponent} from "vue";

export default defineComponent({
  name: 'CommentEntry',
  components: {CommentEditor},
  props: {
    richTextConfig: {
      type: Object,
      default: () => undefined,
    },
    entry: {
      type: Entry,
      required: true,
    },
  },
  emits: ['edit', 'delete'],
  data: function () {
    return {
      isEditing: false,
    };
  },
});
</script>

<style>
.comment-authoring {
  background: #dddddd;
  padding: .5em 1em;
}

.comment-content {
  padding: 1em 0;
}

.comment-actions {
  padding: 4px 0;
}
</style>
