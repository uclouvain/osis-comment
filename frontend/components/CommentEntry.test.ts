/*
 *
 * OSIS stands for Open Student Information System. It's an application
 * designed to manage the core business of higher education institutions,
 * such as universities, faculties, institutes and professional schools.
 * The core business involves the administration of students, teachers,
 * courses, programs and so on.
 *
 * Copyright (C) 2015-2023 UniversitÃ© catholique de Louvain (http://www.uclouvain.be)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * A copy of this license - GNU General Public License - is available
 * at the root of the source code of this program.  If not,
 * see http://www.gnu.org/licenses/.
 *
 */

import {expect, test} from 'vitest';
import {mount} from '@vue/test-utils';
import CommentEntry from "./CommentEntry.vue";
import {Entry} from "../types";

const apiUrl = "/api/dac97c6d-ddb9-47cf-bf72-913fa0ebbbfd/8784bf99-d4f6-4971-99cf-5227afbeca0f";
const comment = "Lorem ipsum dolor amet\nNope";
const entryData = {
  uuid: "8784bf99-d4f6-4971-99cf-5227afbeca0f",
  created_at: "2023-03-02T11:37:05.493008",
  comment: comment,
  author: "John Doe",
  tags: [],
  extra_data: {},
  links: {
    "edit": apiUrl,
    "delete": apiUrl,
  },
};

test('entry', () => {
  expect(CommentEntry).toBeTruthy();

  const wrapper = mount(CommentEntry, {
    props: {entry: new Entry(entryData)},
  });

  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.find('button.btn-default').exists()).toBe(true);
  expect(wrapper.find('button.btn-danger').exists()).toBe(true);
});

test('entry without author', () => {
  const wrapper = mount(CommentEntry, {
    props: {
      entry: new Entry({
        ...entryData,
        author: null,
      }),
    },
  });

  expect(wrapper.find('.comment-authoring strong').text()).toBe('entry.anonymous');
});

test('entry rich text', () => {
  const wrapper = mount(CommentEntry, {
    props: {
      entry: new Entry({
        ...entryData,
        comment: "<p>Lorem ipsum <strong>dolor</strong> amet</p><p>Nope</p>",
      }),
      richTextConfig: {},
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.find('button.btn-default').exists()).toBe(true);
  expect(wrapper.find('button.btn-danger').exists()).toBe(true);
});

test('entry without buttons', () => {
  const wrapper = mount(CommentEntry, {
    props: {
      entry: new Entry({
        ...entryData,
        links: {
          edit: {error: "Nope"},
          delete: {error: "Nope"},
        },
      }),
    },
  });

  expect(wrapper.find('button.btn-default').exists()).toBe(false);
  expect(wrapper.find('button.btn-danger').exists()).toBe(false);
});

test('entry editing', async () => {
  const wrapper = mount(CommentEntry, {
    props: {entry: new Entry(entryData)},
  });

  await wrapper.find('button.btn-default').trigger('click');
  expect(wrapper.findComponent({name: 'CommentEditor'}).exists()).toBe(true);

  await wrapper.find('button.btn-success').trigger('click');
  expect(wrapper.emitted('edit')).toEqual([[apiUrl, comment]]);
});

test('entry editing cancel', async () => {
  const wrapper = mount(CommentEntry, {
    props: {entry: new Entry(entryData)},
  });

  await wrapper.find('button.btn-default').trigger('click');
  expect(wrapper.findComponent({name: 'CommentEditor'}).exists()).toBe(true);

  await wrapper.find('button.btn-default').trigger('click');
  expect(wrapper.findComponent({name: 'CommentEditor'}).exists()).toBe(false);
});

test('entry deletion', async () => {
  const wrapper = mount(CommentEntry, {
    props: {entry: new Entry(entryData)},
  });

  await wrapper.find('button.btn-danger').trigger('click');
  expect(wrapper.emitted('delete')).toEqual([[apiUrl]]);
});
