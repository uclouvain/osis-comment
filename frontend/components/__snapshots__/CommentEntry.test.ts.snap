// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`entry 1`] = `
"<div class=\\"panel-body card-body pb-0 clearfix\\">
  <div class=\\"comment-authoring\\">
    <!--v-if--><strong>John Doe</strong> entry.authored_date
  </div>
  <div class=\\"pull-right float-end comment-actions\\"><button class=\\"btn btn-sm btn-primary\\" title=\\"entry.edit_comment\\"><i class=\\"fas fa-pencil fs-6\\"></i></button><button class=\\"btn btn-sm btn-danger\\" title=\\"entry.delete_comment\\"><i class=\\"fas fa-times fs-6\\"></i></button></div>
  <div class=\\"comment-content pb-0\\">
    <!-- eslint-disable vue/no-v-html -->
    <p>Lorem ipsum dolor amet<br>Nope</p>
  </div>
</div>"
`;

exports[`entry rich text 1`] = `
"<div class=\\"panel-body card-body pb-0 clearfix\\">
  <div class=\\"comment-authoring\\">
    <!--v-if--><strong>John Doe</strong> entry.authored_date
  </div>
  <div class=\\"pull-right float-end comment-actions\\"><button class=\\"btn btn-sm btn-primary\\" title=\\"entry.edit_comment\\"><i class=\\"fas fa-pencil fs-6\\"></i></button><button class=\\"btn btn-sm btn-danger\\" title=\\"entry.delete_comment\\"><i class=\\"fas fa-times fs-6\\"></i></button></div><!-- eslint-disable vue/no-v-html -->
  <div class=\\"comment-content\\">
    <p>Lorem ipsum <strong>dolor</strong> amet</p>
    <p>Nope</p>
  </div>
</div>"
`;
